### PR TITLE
fix: Set childReportID of the optimistic IOU report action

### PR DESCRIPTION
**Pull request by paultsimura**
### Details

When creating a money request, set the report action's `childReportID` to the optimistically created money request `reportID`.

### Fixed Issues

$ https://github.com/Expensify/App/issues/33114
PROPOSAL: https://github.com/Expensify/App/issues/33114#issuecomment-1961778058

### Tests

Same as QA

### Offline tests

Same as QA

#### Test 1:
1. Go offline
2. Request money: repeat for each of the `manual`, `scan`, and `distance` requests;
3. Navigate to the newly created money request report
4. Verify it's showing the correct data
5. Go online
6. Verify the report doesn't disappear after the data is fetched from BE

#### Test 2:
1. Go offline
2. Click "FAB" -> Request money -> Scan
3. Select a receipt, click "next"
4. Next to a user, click "Split"
5. Proceed to create a 1:1 split
6. Click on the created Split preview
7. Enter the amount manually and click "Split"
8. Navigate to the newly created 1:1 money request report
9. Verify it's showing the correct data
10. Go online
11. Verify the report changes to Not Found (this is expected for now because the corresponding BE changes are pending).
12. Navigate back (to the IOU report)
13. Click the IOU Preview again
14. Verify the 1:1 money request report opens and shows the correct data.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/12595293/41d1c124-8186-4ed7-81f7-b38ee095449a

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/12595293/a71d4c72-cff2-4bd8-8ab8-2de79bbdd4cc

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/12595293/fa999c52-20cc-49ff-af8e-432faa571239

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/12595293/cba9bf1a-aae1-4f37-aa12-9e6bbf5efc8a

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/12595293/2a104b9f-1477-4aa6-a506-1a3a2ac4eee4

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/12595293/c3df3194-d540-4721-a826-a1d89d6a074e

</details>

### CODE REVIEW

**Comment by paultsimura** [comment]
I think we can release without it, but there's a BE fix pending: https://github.com/Expensify/App/issues/33114#issuecomment-1961940800

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by situchan** [comment]
Maybe out of scope but there are various glitches as seen in this video:

https://github.com/Expensify/App/assets/108292595/0e694d3a-28c6-4e36-a527-7069b09f0be4

Screenshot 2024-03-04 at 12 01 44 AM
Screenshot 2024-03-04 at 12 01 54 AM
Screenshot 2024-03-04 at 12 02 04 AM
Screenshot 2024-03-04 at 12 02 09 AM
After online:
Screenshot 2024-03-04 at 12 02 41 AM

----

Here's full repro video: (manual request to new (non-Expensify) user in offline mode)

**Comment by situchan** [comment]
BUG: unavailable workspace on transaction detail header when create expense request offline


Screenshot 2024-03-04 at 12 39 18 AM


https://github.com/Expensify/App/assets/108292595/45ede4b1-dc2d-4a15-a591-fbc2457dd2e3

**Comment by situchan** [comment]
@paultsimura this is blocker.

Bug not fixed: split manual request with a single user from FAB


https://github.com/Expensify/App/assets/108292595/7f7307dd-595d-4b92-9c1b-4270f754ee5b

**Comment by paultsimura** [comment]
> @paultsimura this is blocker.
> Bug not fixed: split manual request with a single user from FAB


This one is BE as well, similar to the `SendMoney`.

@cead22 when creating a Split request and calling the `SplitBillAndOpenReport` API endpoint, we send the stringified `splits` among parameters. Each `split` will now contain the `transactionThreadReportID` and `createdReportActionIDForThread` that need to be handled on the BE – currently, the optimistic transaction thread is replaced with another from the BE response.

Here's the `params` example:

```json
{
    "reportID": "1932415687405820",
    "amount": 2200,
    "splits": "[{\"email\":\"paultsimura@gmail.com\",\"accountID\":15528187,\"amount\":1100},{\"email\":\"paultsimura+qa@gmail.com\",\"accountID\":15528255,\"amount\":1100,\"iouReportID\":\"2941200419003018\",\"chatReportID\":\"1932415687405820\",\"transactionID\":\"7451263110636286127\",\"reportActionID\":\"4265162090444479673\",\"createdChatReportActionID\":\"3058872218140896575\",\"createdIOUReportActionID\":\"792942737366932173\",\"reportPreviewReportActionID\":\"7347533532549283559\",\"transactionThreadReportID\":\"5196953162708250\",\"createdReportActionIDForThread\":\"8302969548555128822\"}]",
    "currency": "PLN",
    "merchant": "(none)",
    "created": "2024-03-03 19:54:14.852",
    "comment": "",
    "category": "",
    "tag": "",
    "billable": false,
    "transactionID": "5684861019807192728",
    "reportActionID": "4158501737745037324",
    "policyID": "_FAKE_"
}
```

**Comment by situchan** [comment]
While it's BE issue, I found interesting thing (possible fix in FE).
I tried to create optimistic chat with same user in 2 devices. On 2nd device, after making online, optimistic reportID (wrong one) is automatically replaced with correct reportID.
Can't this be applied to optimistic transaction report as well?


https://github.com/Expensify/App/assets/108292595/fb063372-a58d-4f50-be76-d9146b23c13b

**Comment by paultsimura** [comment]
This looks like a workaround.
The original PR about passing the optimistic thread IDs to BE was intended to fix this very issue, it was just not fully implemented.

**Comment by paultsimura** [comment]
> BUG: unavailable workspace on transaction detail header when create expense manual request offline. Same happens on split manual request with workspace

Found a fix, working on it...

**Comment by paultsimura** [comment]
@situchan I think after this PR this code won't be needed anymore 🤔 

https://github.com/Expensify/App/blob/6c2abde357968c1d38730cd379b3fd654130e766/src/components/ReportActionItem/MoneyRequestAction.tsx#L94-L102

https://github.com/Expensify/App/blob/c2e24e1c8d4ba00c6f206508244224195d0a6591/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L213-L219

**Comment by situchan** [comment]
> @situchan I think after this PR this code won't be needed anymore 🤔 It was used to create a thread if a `moneyRequestPreview` didn't have a child report. But now we are creating and setting it optimistically ever time,
> 
> https://github.com/Expensify/App/blob/6c2abde357968c1d38730cd379b3fd654130e766/src/components/ReportActionItem/MoneyRequestAction.tsx#L94-L102
> 
> https://github.com/Expensify/App/blob/c2e24e1c8d4ba00c6f206508244224195d0a6591/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L213-L219

They were added in #18760, #25429.
Agree cleanup after we make sure that `childReportID` in money request action is already coming from backend for any type of money requests.
1. Login same account on device A & B
2. Create money request in A
3. Navigate to transaction detail page in B

**Comment by paultsimura** [comment]
> * Login same account on device A & B
> * Create money request in A
> * Navigate to transaction detail page in B

Do you observe any bug with this? If so – could you please share the recording? I didn't notice anything suspicious

**Comment by situchan** [comment]
> > * Login same account on device A & B
> > * Create money request in A
> > * Navigate to transaction detail page in B
> 
> Do you observe any bug with this? If so – could you please share the recording? I didn't notice anything suspicious

I just shared test case. If you confirm that no bug found on all types of requests, you can remove them as unnecessary codes. We don't wanna be affected by any possible regressions from code cleanup.

**Comment by situchan** [comment]
Now we should hold PR for BE fix of split, send cases, right?

**Comment by paultsimura** [comment]
> Now we should hold PR for BE fix of split, send cases, right?

I'll let @cead22 decide here. This PR won't break the flow because the behavior we have with splits & send flow is already happening in Prod. But just to be sure the BE fix works well with this PR, we can hold.

**Comment by paultsimura** [comment]
> If you confirm that no bug found on all types of requests, you can remove them as unnecessary codes. We don't wanna be affected by any possible regressions from code cleanup.

Tests well, both IOU & expense requests:

https://github.com/Expensify/App/assets/12595293/78711e40-d237-41e1-a159-75ea9708b398

https://github.com/Expensify/App/assets/12595293/3de721e3-b996-4f18-9a0a-bbee9a86ade5

**Comment by paultsimura** [comment]
> BUG: unavailable workspace on transaction detail header when create expense manual request offline. Same happens on split manual request with workspace

Fixed ✔️

**Comment by cead22** [comment]
> @cead22 when creating a Split request and calling the `SplitBillAndOpenReport` API endpoint, we send the stringified `splits` among parameters. Each `split` will now contain the `transactionThreadReportID` and `createdReportActionIDForThread` that need to be handled on the BE – currently, the optimistic transaction thread is replaced with another from the BE response.

Thanks for letting me know, I'll make an issue for us to update the back end!

**Comment by cead22** [comment]
> currently, the optimistic transaction thread is replaced with another from the BE response.

I made the issue, and we can wait for that fix to test this PR, but are you sure the backend replacing the value with a different one, and the issue isn't that we're not setting `iouAction.childReportID` in the client? 

If the backend isn't returning a bad value, then things should work with the front end fix, because we'll navigate to the right report based on that front end value, right?

**Comment by paultsimura** [comment]
@cead22 I suppose you haven't made any changes on the BE side yet?
That's strange because I cannot reproduce now either `SendMoney` or `Split` bugs (with the changes from this PR) 🤔

**Comment by paultsimura** [comment]
@situchan, from my side the PR now tests well without extra BE changes pending.
Could you please re-test from your side as well?

**Comment by paultsimura** [comment]
> are you sure the backend replacing the value with a different one, and the issue isn't that we're not setting `iouAction.childReportID` in the client?

@cead22, on the recording [here](https://github.com/Expensify/App/pull/37232#issuecomment-1975263308), you can see that the optimistically created transaction thread was present initially (offline), but was replaced with a skeleton once the `ReconnectToReport` API response arrived upon coming back online.

In this PR I'm already setting the `iouAction.childReportID`, and the recording was made with the changes in place, so I wouldn't blame the absent `iouAction.childReportID` here.

Anyway, looks like this blocker was resolved – will just wait for @situchan to confirm.

**Comment by situchan** [comment]
I am blocked retesting [this bug](https://github.com/Expensify/App/pull/37232#issuecomment-1975263308)
Looks like regression from recent PR

https://github.com/Expensify/App/assets/108292595/50f2643a-11b5-4883-a735-9d999854fff5

**Comment by paultsimura** [comment]
> I am blocked retesting [this bug](https://github.com/Expensify/App/pull/37232#issuecomment-1975263308) Looks like regression from recent PR
> 
>  Screen.Recording.2024-03-05.at.8.53.30.PM.mov

I cannot reproduce – what is the precondition? Is it a new user you're trying to split with?

**Comment by situchan** [comment]
Same as https://github.com/Expensify/App/pull/37232#issuecomment-1975263308:

> split manual request with a single user (non-Expensify) from FAB

**Comment by paultsimura** [comment]
@situchan the blocker was resolved, please re-test the failing scenario.
Just make sure to run `npm ci` on the branch – `npm i` did not get rid of the blocker for me.


https://github.com/Expensify/App/assets/12595293/7cb05168-99cc-4fe6-bbda-88ac9988651b

**Comment by situchan** [comment]
Works now after removing node modules again and clearing caches

**Comment by situchan** [comment]
Split scan seems not working. To reproduce, click report preview quickly after scan success


https://github.com/Expensify/App/assets/108292595/005c882e-1d9e-4597-a37e-db4516d368ff

**Comment by situchan** [comment]
Test Result: (🟢: pass, 🔴: failed, 🟡: pending test)

| Request | Manual | Manual (receipt) | Scan | Distance
| :---: | :---: | :---: | :---: | :---:
| IOU (from FAB) |🟢|🟡|🟢|New
| Expense (from FAB) |🟢|🟡|🟢|🟢
| Split with user (from FAB) |🟢|🟡|🔴|New
| Split with workspace (from FAB) |🟡|🟡|🟡|🟡
| Send (from FAB) |🟡|N/A|N/A|N/A
| IOU (from chat) |🟡|🟡|🟡|New
| Expense (from chat) |🟡|🟡|🟡|🟡
| Split with user (from chat) |🟡|🟡|🟡|New
| Split with workspace (from chat) |🟡|🟡|🟡|🟡
| Send (from chat) |🟡|N/A|N/A|N/A


(I keep updating fields upon testing)

**Comment by situchan** [comment]
@paultsimura Let's split testing. Can you please test all above cases for existing contact (if you haven't tested)?
I will test non-existing contact cases.

**Comment by paultsimura** [comment]
@situchan I've confirmed that all manual requests work fine.
Managed reproduce the issue with scan splits, investigating now...

**Comment by paultsimura** [comment]
@cead22 there is a BE issue with the `CompleteSplitBill` API call: it doesn't set (and return) the `childReportID` of the IOU action that we're passing in params. Right after, in an `OpenReport` call to the BE, the `childReportID` comes with a different value. And the optimistically created report gets removed:


https://github.com/Expensify/App/assets/12595293/c2bbf52b-9cef-4037-a2d8-3e9f2dbd84d9

Could we get this fixed on the BE side? Seems like so far it's the only API endpoint that's blocking us (if we want to hold for the BE change).

**Comment by situchan** [comment]
Being many issues are held for this PR and already dragged a lot for past blockers, I suggest to merge without waiting BE fix.

**Comment by situchan** [comment]
Maybe out of scope but map preview not showing in split distance request:

Screenshot 2024-03-07 at 10 31 04 PM

**Comment by situchan** [comment]
Tests completed

Result: https://github.com/Expensify/App/pull/37232#issuecomment-1983648624

**Review by situchan — APPROVED** [review]
All tests pass except split scan.

@cead22 all yours

File: src/components/ReportActionItem/MoneyRequestAction.tsx
Comment (cead22) [inline src/components/ReportActionItem/MoneyRequestAction.tsx:94]: Are we removing this because we're always creating transaction threads in the backend, and `action?.childReportID` will be set to the ID of that transaction thread?

File: src/libs/actions/IOU.ts
Comment (cead22) [inline src/libs/actions/IOU.ts:930]: Should we add a comment above this line saying why we need to do this? It's not super obvious right away, and it made me wonder if we should be setting this in `const iouAction = ReportUtils.buildOptimisticIOUReportAction`, but then realized `ReportUtils.buildTransactionThread` takes `iouAction` which is created by `ReportUtils.buildOptimisticIOUReportAction`.

I don't have a good solution for that right now -- it seems like we could generate an optimistic reportID before the calls to `ReportUtils.buildOptimisticIOUReportAction` and `ReportUtils.buildTransactionThread`, and pass it to both so we don't have to manually set `iouAction.childReportID = optimisticTransactionThread.reportID;`, but I'm not sure if that's a good idea.

I'm curious what you think

File: src/components/ReportActionItem/MoneyRequestAction.tsx
Comment (paultsimura) [reply src/components/ReportActionItem/MoneyRequestAction.tsx:94]: Yes. We've discussed it briefly here: https://github.com/Expensify/App/pull/37232#issuecomment-1975285144

File: src/libs/actions/IOU.ts
Comment (paultsimura) [reply src/libs/actions/IOU.ts:930]: Yes, unfortunately these 2 are co-dependent: the transaction thread has the iou action's ID as the parent action ID, and vice versa. Also, passing the pre-generated report ID will cause a lot of refactoring of the current methods that use the `buildOptimisticChatReport`. One possible idea here is to make a separate function that will generate the `iouAction`, `optimisticTransactionThread`, and the `optimisticCreatedActionForTransactionThread`. This way, we'll not be setting the `iouAction.childReportID` in 4 different places, but in only one function, and will be calling it from those 4 places. It's not perfect either, but will minimize code duplication.

Smth like:

```js
function buildOptimisticIOUReportActionWithTransactionThread(...) {
  const iouAction = ReportUtils.buildOptimisticIOUReportAction(
        CONST.IOU.REPORT_ACTION_TYPE.CREATE,
        amount,
        currency,
        comment,
        [participant],
        optimisticTransaction.transactionID,
        undefined,
        iouReport.reportID,
        false,
        false,
        receiptObject,
        false,
        currentTime,
    );
    const optimisticTransactionThread = ReportUtils.buildTransactionThread(iouAction, iouReport.reportID);
    iouAction.childReportID = optimisticTransactionThread.reportID;
    const optimisticCreatedActionForTransactionThread = ReportUtils.buildOptimisticCreatedReportAction(payeeEmail);


  return { iouAction, optimisticTransactionThread, optimisticCreatedActionForTransactionThread }
}
```

Comment (cead22) [reply src/libs/actions/IOU.ts:930]: Yeah I thought of that but wasn't convinced it would be better. But on second thought, I do think it's better and we can add a comment inside that function above where we set `iouAction.childReportID` so people know not to remove it.

The alternative would be to put that comment in 4 places, and we'd have that code duplicated like you said, and nothing prevents people from forgetting to set `iouAction.childReportID`

I'm good to update to this 👍

Comment (paultsimura) [reply src/libs/actions/IOU.ts]: Done ✔️ 

@situchan I've re-tested the 4 different flows that use the 4 different functions that call this new `buildOptimisticMoneyRequestEntities` function – everything works as expected (except for the known Complete Split BE issue):
- Send Money
- Request Money
- Split right away (the fixed amount split)
- Split with Complete Split (the scan split)

**Comment by paultsimura** [comment]
@cead22 while testing, I've noticed a (relatively unrelated) issue with the `SendMoney` flow – we don't properly pass the optimistic `CREATED` actions:

https://github.com/Expensify/App/assets/12595293/47e059c1-62ec-444b-93e8-a53e7eb0a0c1

But to fix it, we need another BE fix. Should we hold for it, or can we make it a separate issue?

**Comment by situchan** [comment]
> @cead22 while testing, I've noticed a (relatively unrelated) issue with the `SendMoney` flow – we don't properly pass the optimistic `CREATED` actions:
> 
>  Screen.Recording.2024-03-10.at.19.12.42-compressed.mp4 
> But to fix it, we need another BE fix. Should we hold for it, or can we make it a separate issue?

yes, I noticed that as well but that should not really block this PR to be merged

**Review by cead22 — APPROVED** [review]

**Comment by cead22** [comment]
All checks passed and everything was green when I merged. This wasn't an emergency change
image

**Pull request merged**
